pkg/retriever/httpretriever.go

```diff
@@ -6,255 +6,164 @@ import (
-			if !ok {
-				// we failed, and got only retrieval errors
-				retrievalErrors = multierr.Append(retrievalErrors, ErrAllRetrievalsFailed)
-				return nil, retrievalErrors
-			}
+func (ph ProtocolHttp) CompareCandidates(a, b connectCandidate, mda, mdb metadata.Protocol) bool {
+	// we only have duration .. currently
+	return a.Duration < b.Duration
+}
 
-			if result.Event != nil {
-				r.eventsCallback(*result.Event)
-				break
-			}
-			if result.Err != nil {
-				retrievalErrors = multierr.Append(retrievalErrors, result.Err)
-			}
-			if result.Stats != nil {
-				return result.Stats, nil
-			}
-		case <-ctx.Done():
-			return nil, context.Canceled
-		}
+func (ph *ProtocolHttp) Connect(ctx context.Context, retrieval *retrieval, candidate types.RetrievalCandidate) error {
+	// we currently start a full HTTP request during "connect" phase, which means
+	// we'll hit all candidates in parallel before proceeding to read the body
+	// of ones we choose, one by one, until we get success. This may not be
+	// optimal for servers that have to queue with a body, and may also result in
+	// timeouts for body reading when we fail on one and move to another.
+	// This may all need to move into the Retrieve() call and Connect() be a noop.
+	var err error
+	ph.req, err = makeRequest(ctx, retrieval.request, candidate)
+	if err == nil {
+		ph.resp, err = ph.Client.Do(ph.req)
 	}
+	return err
 }
 
-func (r *httpRetrieval) doHTTPDownload(
+func (ph *ProtocolHttp) Retrieve(
 	ctx context.Context,
+	retrieval *retrieval,
+	session *retrievalSession,
 	phaseStartTime time.Time,
+	timeout time.Duration,
 	candidate types.RetrievalCandidate,
-) {
-	var stats *types.RetrievalStats
-	var retrievalErr error
+) (*types.RetrievalStats, error) {
 
-	r.sendEvent(events.Started(r.HTTPRetriever.Clock.Now(), r.request.RetrievalID, phaseStartTime, types.RetrievalPhase, candidate))
+	defer ph.resp.Body.Close()
+	return readBody(candidate.RootCid, candidate.MinerPeer.ID, ph.resp.Body, retrieval.request.LinkSystem.StorageWriteOpener)
+}
 
-	var candidateURL *url.URL
-	var err error
-	for _, addr := range candidate.MinerPeer.Addrs {
-		candidateURL, err = maurl.ToURL(addr)
-		if err == nil {
-			break
-		}
-	}
+func makeRequest(ctx context.Context, request types.RetrievalRequest, candidate types.RetrievalCandidate) (*http.Request, error) {
+	candidateURL, err := candidate.ToURL()
 	if err != nil {
 		log.Warnf("Couldn't construct a url for miner %s: %v", candidate.MinerPeer.ID, err)
-		retrievalErr = fmt.Errorf("%w: %v", ErrConnectFailed, err)
-		r.sendEvent(events.Failed(r.HTTPRetriever.Clock.Now(), r.request.RetrievalID, phaseStartTime, types.RetrievalPhase, candidate, retrievalErr.Error()))
-		return
+		return nil, fmt.Errorf("%w: %v", ErrNoHttpForPeer, err)
 	}
 
-	reqURL := fmt.Sprintf("%s/ipfs/%s%s", candidateURL, r.request.Cid, r.request.Path)
-	req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
+	path, err := request.GetUrlPath()
```

Comment:
this is certainly a 'hack' and we maybe should leave at least a todo/issue here

I think the path to being able to schedule multiple peers involving graphsync & http involves:
* being able to construct the path / request here from the selector of the candidate
* being able to cancel individual graphsync / http requests early
* being able to schedule subsequent requests for later in the traversal from other peers once we identify those opportunities.

Reply:
I don't think this particular bit is a hack per se. Perhaps looking at it in the diff view makes it seem that way? The diff makes a complete mess of what's going on here but it's actually straightforward and does make the full request based on the incoming candidate and original lassie request.

The only unfortunate part of this bit of code is the fact that we have to error on non-path requests - so if you show up with a custom selector then it'll bail at this point. So space is open here to branch into a different form and do a manual traversal, block-by-block via http, which isn't going to be pretty, but possible at least and we have similar code in the bitswap retriever to do the same thing.

The tricky stuff about cancelling early and being able to come back to revisit parts of a missing DAG are a challenge for all our retrievers.

Right now the status quo for cancellation & racing is:

* We try all protocols we have valid and acceptable candidates for; racing them in parallel
* We do per-protocol scheduling:
  - Bitswap is just a pool of candidates, no need to schedule
  - Graphsync & HTTP do this dance of parallel init but an actual data fetch of one candidate at a time until they get to success. Once they get a success, they cancel anything left in the queue.
* When we get a successful response from any protocol, we cancel the others, regardless of where they're up to.

All of this works now with HTTP too, albeit with the questions about what "init" means or should mean for HTTP. (To be resolved later probably).

Because we share a LinkSystem between them all, traversals should be handling the parallelism already. There's going to be some wasted transfer of blocks (e.g. Graphsync server sends blocks that it thinks we don't have but it turns out we suddenly have them coming in from Bitswap). But the resulting CAR shouldn't contain duplicates.

Lots of opportunities all over there for smarter coordination between protocols. But also lots of hard challenges.

Reply:
fair enough - i think i looked at that and assumed the `Scope` and `Path` were coming directly from the http request itself and weren't as integrated as they actually are. Things are also a bit artificially simplified by not dealing with range requests yet.